### PR TITLE
unify empty_cpu and empty_cuda, call native::empty where appropriate

### PR DESCRIPTION
**Pull request by ngimel**
per title

### CODE REVIEW

File: aten/src/ATen/native/TensorIterator.cpp [hunk 1]
```diff
@@ -6,6 +6,7 @@
 #include <ATen/native/TypeProperties.h>
 #include <ATen/MemoryOverlap.h>
 #include <ATen/native/Resize.h>
+#include <ATen/core/grad_mode.h>
```
Comment (mcarilli) [inline aten/src/ATen/native/TensorIterator.cpp:9]: Using `AutoNonVariableTypeMode` here is a good thing (I think) but I don't think you need `grad_mode.h` for `AutoNonVariableTypeMode`. `GradMode` is a standalone thread-local flag, a temporary hack separate from the the dispatch hierarchy.  `AutoNonVariableTypeMode` controls the true autograd dispatch keys, which determine whether an op dispatches through `VariableType*` at all.  `GradMode` is a separate flag that determines if grad history recording takes place if you happen to route through `VariableType*`.

While working on the torchvision PR I beat my head through the [distinction between the two](https://github.com/pytorch/pytorch/issues/40736#issuecomment-652129302).  my comment is almost two months old, but given that the [TODO to unify GradMode and the autograd dispatch keys](https://github.com/pytorch/pytorch/blob/7f967c08b8ae47bd400df53caafa5061e9bc3bb4/aten/src/ATen/core/LegacyTypeDispatch.h#L39-L44) in `LegacyTypeDispatch.h` is still present, the fact that autograd dispatch selection [still leaves DispatchKey::Autograd always-on-by-default](https://github.com/pytorch/pytorch/blob/7f967c08b8ae47bd400df53caafa5061e9bc3bb4/aten/src/ATen/core/dispatch/DispatchKeyExtractor.h#L18), and the fact that GradMode is still present in Pytorch at all, my comment is not obsolete.  @ezyang would know more.

File: aten/src/ATen/native/TensorIterator.cpp [hunk 2]
```diff
@@ -402,9 +402,11 @@ void TensorIterator::compute_types(const TensorIteratorConfig& config) {
       // Casts to outputs by creating temporaries of the correct dtype (if needed)
       if (config.cast_common_dtype_to_outputs_ && op.is_output && op.current_dtype != common_dtype_) {
         op.original_tensor = op.tensor;
-        op.tensor = at::empty_like(op.tensor,
-                                   op.tensor.options().dtype(common_dtype_),
-                                   LEGACY_CONTIGUOUS_MEMORY_FORMAT);
+        AutoNonVariableTypeMode non_variable_type_mode;
+        op.tensor = at::native::empty_like(
```
Comment (mcarilli) [inline aten/src/ATen/native/TensorIterator.cpp:406]: Do we care this no longer gets tracing?  Imo we don't care:  the performance gain of dodging dispatch altogether for `at::native::empty*` is worth missing tracing.  Tracing will still profile the overarching TensorIterator op.  If tracing says some TensorIterator op has weirdly high cpu overhead, we can switch to a focused flame graph (which we should do even if `empty*` gave tracing info, and in such case one of the first things the flame graph would tell us would be:  get rid of dispatch+tracing :P).

Comment (ngimel) [reply aten/src/ATen/native/TensorIterator.cpp:406]: We did not trace it for the majority of history, we started tracing it a few months back, and we can stop IMO. I agree that it's not providing a lot of value.

File: aten/src/ATen/native/TensorIterator.cpp [hunk 3]
```diff
@@ -1087,7 +1091,9 @@ bool TensorIterator::fast_set_up(const TensorIteratorConfig& config) {
           auto& op = operands_[i];
           if (!op.tensor.defined()) {
             TORCH_INTERNAL_ASSERT(op.is_type_defined(), "no type for operand", i);
-            op.tensor = at::empty(shape_, op.options(), MemoryFormat::Contiguous);
+            AutoNonVariableTypeMode non_variable_type_mode;
+            const DeviceGuard device_guard(op.options().device());
```
Comment (ngimel) [inline aten/src/ATen/native/TensorIterator.cpp:1095]: NB: DeviceGuard is not needed here, it's needed only for scalar-tensor case in non-fast setup, remove it later.

**Review by bhosmer — COMMENTED** [review]
Subtleties aside, I'm mostly interested in understanding whether this represents a newly explicit restriction on TensorIterator usage, and if so, if that warrants additional checking (per inline comments).

**Comment by ngimel** [comment]
> Subtleties aside, I'm mostly interested in understanding whether this represents a newly explicit restriction on TensorIterator usage, and if so, if that warrants additional checking (per inline comments).

That's a very good question, and the answer is slightly complicated. Currently, only cpu, cuda and quantized use TensorIterator. I'd argue that since most of the value of TensorIterator lies in facilitating operations on tensors with arbitrary strides (and TensorIterator explicitly accesses strides in many places), only tensors with strided layouts with explicit strides should be using it. So, e.g. at::empty call could be dispatched to mkl tensor and sparse tensor, but mkl and sparse never should use TensorIterator, because in mkl's case the layout is opaque, and in sparse it's completely different. In that sense, replacing at::empty with at::native::empty_cpu_or_cuda is not narrowing the potential usages. 
Now, to quantized. Even before this PR, at::empty could not be dispatched to quantized tensors and would error out. That's why all quantization usages were very careful to allocate their outputs before, so that they never hit those paths, so quantized usage is unchanged. 
So tl;dr it does not break anything in the current codebase, it would if we were to add another strided type with `empty` dispatch, or if an extension defined such a type and wanted to use TensorIterator on it.

File: aten/src/ATen/native/TensorFactories.cpp [hunk 4]
```diff
@@ -196,16 +201,19 @@ Tensor empty_cpu(IntArrayRef size, const TensorOptions& options_, c10::optional<
...
     tensor.unsafeGetTensorImpl()->set_sizes_contiguous(size);
   }
 
   auto memory_format = options.memory_format_opt().value_or(MemoryFormat::Contiguous);
-  tensor.unsafeGetTensorImpl()->empty_tensor_restride(memory_format);
-
+  //for contiguous format restride was done in set_sizes_contiguous, restride cases that did not go there
+  if ((size.size()==1 && size[0]==0) || memory_format != MemoryFormat::Contiguous) {
+      tensor.unsafeGetTensorImpl()->empty_tensor_restride(memory_format);
+  }
```
Comment (ezyang) [inline aten/src/ATen/native/TensorFactories.cpp:216]: Oof, this is so ugly

File: aten/src/ATen/native/cuda/TensorFactories.cu [hunk 5]
```diff
@@ -43,41 +43,8 @@ Tensor& eye_out_cuda(Tensor& result, int64_t n, int64_t m) {
   return result;
 }
 
-Tensor empty_cuda(IntArrayRef size, const TensorOptions& options, c10::optional<MemoryFormat> optional_memory_format) {
```
Comment (ezyang) [inline aten/src/ATen/native/cuda/TensorFactories.cu:46]: Thank you, good riddance.

File: aten/src/ATen/native/TensorIterator.cpp [hunk 6]
```diff
@@ -402,9 +402,11 @@ void TensorIterator::compute_types(const TensorIteratorConfig& config) {
       // Casts to outputs by creating temporaries of the correct dtype (if needed)
       if (config.cast_common_dtype_to_outputs_ && op.is_output && op.current_dtype != common_dtype_) {
         op.original_tensor = op.tensor;
-        op.tensor = at::empty_like(op.tensor,
-                                   op.tensor.options().dtype(common_dtype_),
-                                   LEGACY_CONTIGUOUS_MEMORY_FORMAT);
+        AutoNonVariableTypeMode non_variable_type_mode;
```
Comment (ezyang) [inline aten/src/ATen/native/TensorIterator.cpp:405]: The AutoNonVariableTypeMode additions here are super sketchy. This is related to the Q you asked earlier in composability chat:

> I've submitted https://github.com/pytorch/pytorch/pull/43773 where I make TensorIterator use at::native::empty instead of at::empty, to avoid dispatch costs. I had test failures related to autograd tests for logit function. logit derivative is defined as 
> - name: logit(Tensor self, float? eps=None) -> Tensor
  self: "GradMode::is_enabled() ? infinitely_differentiable_logit_backward(grad, self, eps) : logit_backward(grad, self, eps)"
> 
> and in logit_backward I was getting 
> RuntimeError: impl::variable_excluded_from_dispatch() INTERNAL ASSERT FAILED at "../aten/src/ATen/native/TensorFactories.cpp":182, please report a bug to PyTorch. 
Exception raised from empty at ../aten/src/ATen/native/TensorFactories.cpp:182 (most recent call first):
>
> I worked around it by adding not_implemented gradients for logit_backward function. It's still weird that I had to do that, because logit_backward is called only when GradMode is not enabled. Interestingly, silu and prelu were not affected and did not require this workaround, even though their derivatives are similar, and gradients for silu_backward are not defined:
> - name: silu(Tensor self) -> Tensor
>   self: "GradMode::is_enabled() ? infinitely_differentiable_silu_backward(grad, self) : silu_backward(grad, self)"
>
> Any idea why that would be? One thing that's different between them is logit is exposed as torch function and method, silu is accessible only via torch.nn.functional. 
> My PR brings ~10% speed-up for out-of-place operations overheads 

I'd definitely like to get a good answer about why we are calling TensorIterator without autograd mode having already been disabled; my preference would have been for us to fix those cases instead of directly bashing them in TensorIterator itself.

**Comment by ezyang** [comment]
This is nice work. I'd like to understand a little better what the state of the universe is (since the business with AutoNonVariableTypeMode is a bit sketchy) and I would also like to understand where the 10% performance gain is coming from. But 10% is nothing to sniff at, and I definitely want to see this go in.

File: aten/src/ATen/native/TensorIterator.cpp (hunk 6 above)
Comment (ngimel) [reply aten/src/ATen/native/TensorIterator.cpp:405]: This is happening in the tests https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/test/tensor_iterator_test.cpp#L54-L56. I'm guessing it also could happen if someone were to use TensorIterator in custom op. 
To add to this, DeviceGuard business is also sketchy, we should be calling TensorIterator with proper deviceGuard already, and yet in scalar-tensor case we don't (long story). And again, same thing can happen with people writing custom ops, they won't have properly set device guards.

File: aten/src/ATen/native/TensorFactories.cpp (hunk 4 above)
Comment (ngimel) [reply aten/src/ATen/native/TensorFactories.cpp:216]: Agreed, I'll see if I can get rid of `size` part of this conditional.

**Comment by bhosmer** [comment]
> So tl;dr it does not break anything in the current codebase, it would if we were to add another strided type with `empty` dispatch, or if an extension defined such a type and wanted to use TensorIterator on it.

Ah, thanks for all the detail! Given that rationale for when TensorIterator should be used, is it worth adding an explicit check in TensorIteratorConfig? Just eyeballing [what seems like a likely spot](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/TensorIterator.cpp#L293-L296) it doesn't look like a check would add substantially to the current overhead, and it might usefully solidify the contract that (per your description) is already implicitly in place.

Anyway though, this is a minor item compared to what you and Ed are discussing above. If there are issues in adding such a check that I'm not aware of, or if you don't want to muddy the waters in this PR or whatever, just disregard 😁

File: aten/src/ATen/native/TensorIterator.cpp (hunk 6 above)
Comment (ezyang) [reply aten/src/ATen/native/TensorIterator.cpp:405]: Oh ok, this makes a good deal of sense.

For the custom op situation, @bhosmer and @ailzhang recently agreed with me that in the long term we should have a fallback autograd kernel which excludes all dispatch keys, which would make custom ops work correctly. I suppose you are right that without these calls, in the short term you would break all custom ops using tensor iterator without these guards.

So OK, I'm fine with these guards, but I would appreciate it if we had a comment saying why the guards are needed and under what circumstances they could be removed.
